torchinductor/codegen/jiterator.py

```diff
@@ -0,0 +1,288 @@
+        name = str(name)
+        if name not in odict:
+            odict[name] = f"{prefix}{len(odict)}"
+        return odict[name]
+
+    def __init__(self, sizevars=None):
+        self.input_buffers = collections.OrderedDict()
+        self.output_buffers = collections.OrderedDict()
+        self.sizevars = sizevars or collections.OrderedDict()
+
+    def input(self, name):
+        assert name not in V.graph.removed_buffers, name
+        if name in self.output_buffers:
+            return self.output_buffers[name]
+        return self._lookup("in", self.input_buffers, name)
+
+    def output(self, name):
+        assert name not in V.graph.removed_buffers, name
+        assert name not in self.input_buffers, name
+        return self._lookup("out", self.output_buffers, name)
+
+    def size(self, name):
+        return self._lookup("ks", self.sizevars, name)
+
+    def call_names(self):
+        return chain(
+            self.input_buffers.keys(), self.output_buffers.keys(), self.sizevars.keys()
+        )
+
+    def argdefs(self):
+        # from .cpp import DTYPE_TO_CPP
+        # from .cpp import INDEX_TYPE
+        argdefs = []
+        for outer, inner in self.input_buffers.items():
+            # dtype = buffer_types[outer]
+            # argdefs.append(f"const {DTYPE_TO_CPP[dtype]} {inner}")
+            argdefs.append(f"const T {inner}")
+        return argdefs
+
+class KernelGroup:
+    def __init__(self):
+        super().__init__()
+        self.args = JiteratorKernelArgs()
+        self.loops_code = BracesBuffer()
+        self.stack = contextlib.ExitStack()
+        self.count = 0
+
+    def new_kernel(self):
+        return JiteratorKernel(self.args)
+
+    def finalize_kernel(self, new_kernel, scheduler):
+        self.count += 1
+        code = self.loops_code
+        new_kernel.codegen_loops(code)
+
+        if not scheduler.runable_nodes and scheduler.pending_buffer_names:
+            scheduler.barrier()
+
+    def codegen_define_and_call(self, wrapper):
+        self.stack.close()
+        if self.count == 0:
+            return
+
+        argdefs = ",\n".ljust(25).join(self.args.argdefs())
+        code = BracesBuffer()
+        code.writelines([f'template <typename T> T kernel({argdefs})'])
+
+        with code.indent():
+            code.splice(self.loops_code)
+
+        codecache_def = IndentedBuffer()
+        codecache_def.writeline("'''")
+        codecache_def.splice(code)
+        codecache_def.writeline("'''")
+
+        kernel_name = wrapper.next_kernel_name()
+        wrapper.define_kernel(kernel_name, codecache_def.getvalue())
+
+        # generate the code to call this
+        args = self.args
+        call_args = []
+        for name in args.input_buffers.keys():
+            call_args.append(f"{name}")
+
+        output_args = []
+        for name in args.output_buffers.keys():
+            output_args.append(f"{name}")
+
+        wrapper.body.writeline(
```

Comment:
we may want to write this to the header section so it gets compiled head of time.

Reply:
Yes. I moved the _create_jit_fn declaration to the header. 

Under the hook, the compilation actually happens during the first call. Future calls will read the compiled kernel from a Jiterator internal cache.